File: cpp/src/0012.cpp
/*
[12. Integer to Roman (Medium)](https://leetcode.com/problems/integer-to-roman/)

@topic: Hash Table
@topic: Math
@topic: String

Constraints:
  - `1 <= num <= 3999`
*/

#include <string>

using std::string;

class Solution {
   public:
    string intToRoman(int num) {
        const static string thousands[] = {"", "M", "MM", "MMM"};
        const static string hundreds[] = {"", "C", "CC", "CCC", "CD", "D", "DC", "DCC", "DCCC", "CM"};
        const static string tens[] = {"", "X", "XX", "XXX", "XL", "L", "LX", "LXX", "LXXX", "XC"};
        const static string ones[] = {"", "I", "II", "III", "IV", "V", "VI", "VII", "VIII", "IX"};
        return thousands[num / 1000] + hundreds[num % 1000 / 100] + tens[num % 100 / 10] + ones[num % 10];
    }
};

#include <cassert>
#include <chrono>
#include <iostream>
#include <vector>

using std::tuple;
using std::vector;
using std::chrono::duration_cast;
using std::chrono::microseconds;
using std::chrono::system_clock;

int main() {
    Solution o;

    vector<tuple<int, string>> CASES = {
        {3, "III"},
        {58, "LVIII"},
        {1994, "MCMXCIV"},
    };

    for (auto& [num, excepted] : CASES) {
        assert(o.intToRoman(num) == excepted);
    }

    auto start = system_clock::now();
    for (auto& [num, _] : CASES) {
        for (int i = 0; i < 100000; ++i) {
            o.intToRoman(num);
        }
    }
    auto end = system_clock::now();
    std::cout << duration_cast<microseconds>(end - start).count() << std::endl;
}

File: csharp/src/0012.cs
using System;
using System.Diagnostics;

public class Solution {
    public string IntToRoman(int num) {
        return (
            (num / 1000) switch {
                1 => "M",
                2 => "MM",
                3 => "MMM",
                _ => "",
            } +
            (num % 1000 / 100) switch {
                1 => "C",
                2 => "CC",
                3 => "CCC",
                4 => "CD",
                5 => "D",
                6 => "DC",
                7 => "DCC",
                8 => "DCCC",
                9 => "CM",
                _ => "",
            } +
            (num % 100 / 10) switch {
                1 => "X",
                2 => "XX",
                3 => "XXX",
                4 => "XL",
                5 => "L",
                6 => "LX",
                7 => "LXX",
                8 => "LXXX",
                9 => "XC",
                _ => "",
            } +
            (num % 10) switch {
                1 => "I",
                2 => "II",
                3 => "III",
                4 => "IV",
                5 => "V",
                6 => "VI",
                7 => "VII",
                8 => "VIII",
                9 => "IX",
                _ => "",
            });
    }

    static void Main(string[] args) {
        var o = new Solution();

        Debug.Assert(o.IntToRoman(3) == "III");
        Debug.Assert(o.IntToRoman(58) == "LVIII");
        Debug.Assert(o.IntToRoman(1994) == "MCMXCIV");

        var timer = new Stopwatch();
        timer.Start();
        for (var i = 0; i < 1000000; i++) {
            o.IntToRoman(3);
            o.IntToRoman(58);
            o.IntToRoman(1994);
        }
        timer.Stop();
        Console.WriteLine(timer.Elapsed.TotalMilliseconds);
    }
}
